man/in2csv.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "IN2CSV" "1" "Aug 16, 2024" "2.2.0" "csvkit"
.SH NAME
in2csv \- in2csv Documentation
.SH DESCRIPTION
.sp
Converts various tabular data formats into CSV.
.sp
Converting fixed width requires that you provide a schema file with the \(dq\-s\(dq option. The schema file should have the following format:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
column,start,length
name,0,30
birthday,30,10
age,40,3
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The header line is required though the columns may be in any order:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: in2csv [\-h] [\-d DELIMITER] [\-t] [\-q QUOTECHAR] [\-u {0,1,2,3}] [\-b]
              [\-p ESCAPECHAR] [\-z FIELD_SIZE_LIMIT] [\-e ENCODING] [\-L LOCALE]
              [\-S] [\-\-blanks] [\-\-null\-value NULL_VALUES [NULL_VALUES ...]]
              [\-\-date\-format DATE_FORMAT] [\-\-datetime\-format DATETIME_FORMAT]
              [\-H] [\-K SKIP_LINES] [\-v] [\-l] [\-\-zero] [\-V]
              [\-f {csv,dbf,fixed,geojson,json,ndjson,xls,xlsx}] [\-s SCHEMA]
              [\-k KEY] [\-n] [\-\-sheet SHEET] [\-\-write\-sheets WRITE_SHEETS]
              [\-\-use\-sheet\-names] [\-\-reset\-dimensions]
              [\-\-encoding\-xls ENCODING_XLS] [\-y SNIFF_LIMIT] [\-I]
              [FILE]

Convert common, but less awesome, tabular data formats to CSV.

positional arguments:
  FILE                  The CSV file to operate on. If omitted, will accept
                        input as piped data via STDIN.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-f {csv,dbf,fixed,geojson,json,ndjson,xls,xlsx}, \-\-format {csv,dbf,fixed,geojson,json,ndjson,xls,xlsx}
                        The format of the input file. If not specified will be
                        inferred from the file type.
  \-s SCHEMA, \-\-schema SCHEMA
                        Specify a CSV\-formatted schema file for converting
                        fixed\-width files. See web documentation.
  \-k KEY, \-\-key KEY     Specify a top\-level key to look within for a list of
                        objects to be converted when processing JSON.
  \-n, \-\-names           Display sheet names from the input Excel file.
  \-\-sheet SHEET         The name of the Excel sheet to operate on.
  \-\-write\-sheets WRITE_SHEETS
                        The names of the Excel sheets to write to files, or
                        \(dq\-\(dq to write all sheets.
  \-\-use\-sheet\-names     Use the sheet names as file names when \-\-write\-sheets
                        is set.
  \-\-reset\-dimensions    Ignore the sheet dimensions provided by the XLSX file.
  \-\-encoding\-xls ENCODING_XLS
                        Specify the encoding of the input XLS file.
  \-y SNIFF_LIMIT, \-\-snifflimit SNIFF_LIMIT
                        Limit CSV dialect sniffing to the specified number of
                        bytes. Specify \(dq0\(dq to disable sniffing entirely, or
                        \(dq\-1\(dq to sniff the entire file.
  \-I, \-\-no\-inference    Disable type inference (and \-\-locale, \-\-date\-format,
                        \-\-datetime\-format, \-\-no\-leading\-zeroes) when parsing
                        CSV input.

 Some command\-line flags only pertain to specific input formats.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See also: \fI\%Arguments common to all tools\fP\&.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
The \(dqndjson\(dq format refers to \(dqnewline delimited JSON\(dq, as used by many streaming APIs.
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
If an XLS looks identical to an XLSX when viewed in Excel, they may not be identical as CSV. For example, XLSX has an integer type, but XLS doesn\(aqt. Numbers that look like integers from an XLS will have decimals in CSV, but those from an XLSX won\(aqt.
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
To convert from HTML, consider \fI\%messytables\fP\&.
.UNINDENT
.UNINDENT
.SH EXAMPLES
.sp
Convert the 2000 census geo headers file from fixed\-width to CSV and from latin\-1 encoding to utf8:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv \-e iso\-8859\-1 \-f fixed \-s examples/realdata/census_2000/census2000_geo_schema.csv examples/realdata/census_2000/usgeo_excerpt.upl
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
A library of fixed\-width schemas is maintained in the \fBffs\fP project:
.sp
\fI\%https://github.com/wireservice/ffs\fP
.UNINDENT
.UNINDENT
.sp
Convert an Excel .xls file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv examples/test.xls
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Standardize the formatting of a CSV file (quoting, line endings, etc.):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv examples/realdata/FY09_EDU_Recipients_by_State.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Fetch csvkit\(aqs open issues from the GitHub API, convert the JSON response into a CSV and write it to a file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
curl https://api.github.com/repos/wireservice/csvkit/issues?state=open | in2csv \-f json \-v
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Convert a DBase DBF file to an equivalent CSV:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv examples/testdbf.dbf
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This tool names unnamed headers. To avoid that behavior, run:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv \-\-no\-header\-row examples/test.xlsx | tail \-n +2
.ft P
.fi
.UNINDENT
.UNINDENT
.SH TROUBLESHOOTING
.sp
If an error like the following occurs when providing an input file in CSV or Excel format:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
ValueError: Row 0 has 11 values, but Table only has 1 columns.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Then the input file might have initial rows before the header and data rows. You can skip such rows with \fB\-\-skip\-lines\fP (\fB\-K\fP):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
in2csv \-\-skip\-lines 3 examples/test_skip_lines.csv
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If an XLSX file yields too few rows or too few columns, then the application that created the file might have \fI\%incorrectly set the worksheet\(aqs dimensions\fP\&. Try again with the \fB\-\-reset\-dimensions\fP option.
.SH AUTHOR
Christopher Groskopf and contributors
.SH COPYRIGHT
2016, Christopher Groskopf and James McKinney
.\" Generated by docutils manpage writer.
.


CHANGELOG
0.9.2
-----

* Fix version of openpyxl.

0.9.1
-----

* Add Antonio Lima to AUTHORS.
* Add support for ndjson. (#329)
* Add missing docs for csvcut -C. (#227)
* Reorganize docs so TOC works better. (#339)
* Render docs locally with RTD theme.
* Fix header in "tricks" docs.
* Add install instructions to tutorial. (#331)
* Add killer examples to doc index. (#328)
* Reorganize doc index
* Fix broken csvkit module documentation. (#327)
* Fix version of openpyxl to work around encoding issue. (#391, #288)

0.9.0
-----

* Write missing sections of the tutorial. (#32)
* Remove -q arg from sql2csv (conflicts with common flag).
* Fix csvjoin in case where left dataset rows without all columns.
* Rewrote tutorial based on LESO data. (#324)
* Don't error in csvjson if lat/lon columns are null. (#326)
* Maintain field order in output of csvjson.
* Add unit test for json in2csv. (#77)
* Maintain key order when converting JSON into CSV. (#325.)
* Upgrade python-dateutil to version 2.2 (#304)
* Fix sorting of columns with null values. (#302)
* Added release documentation.
* Fill out short rows with null values. (#313)
* Fix unicode output for csvlook and csvstat. (#315)
* Add documentation for --zero. (#323)
* Fix Integrity error when inserting zero rows in database with csvsql. (#299)
* Add Michael Mior to AUTHORS. (#305)
* Add --count option to CSVStat.
* Implement csvformat.
* Fix bug causing CSVKitDictWriter to output 'utf-8' for blank fields.

0.8.0
-----

* Add pnaimoli to AUTHORS.
* Fix column specification in csvstat. (#236)
* Added "Tips and Tricks" documentation. (#297, #298)
* Add Espartaco Palma to AUTHORS.
* Remove unnecessary enumerate calls. (#292)
* Deprecated DBF support for Python 3+.
* Add support for Python 3.3 and 3.4 (#239)

0.7.3
-----

* Fix date handling with openpyxl > 2.0 (#285)
* Add Kristina Durivage to AUTHORS. (#243)
* Added Richard Low to AUTHORS.
* Support SQL queries "directly" on CSV files. (#276)
* Add Tasneem Raja to AUTHORS.
* Fix off-by-one error in open ended column ranges. (#238)
* Add Matt Pettis to AUTHORS.
* Add line numbers flag to csvlook (#244)
* Only install argparse for Python < 2.7. (#224)
* Add Diego Rabatone Oliveira to AUTHORS.
* Add Ryan Murphy to AUTHORS.
* Fix DBF dependency. (#270)

0.7.2
-----

* Fix CHANGELOG for release.

0.7.1
-----

* Fix homepage url in setup.py.

0.7.0
-----

* Fix XLSX datetime normalization bug. (#223)
* Add raistlin7447 to AUTHORS.
* Merged sql2csv utility (#259).
* Add Jeroen Janssens to AUTHORS.
* Validate csvsql DB connections before parsing CSVs. (#257)
* Clarify install process for Ubuntu. (#249)
* Clarify docs for --escapechar. (#242)
* Make ``import csvkit`` API compatible with ``import csv``.
* Update Travis CI link. (#258)
* Add Sébastien Fievet to AUTHORS.
* Use case-sensitive name for SQLAlchemy (#237)
* Add Travis Swicegood to AUTHORS.

0.6.1
-----

* Add Chris Rosenthal to AUTHORS.
* Fix multi-file input to csvsql. (#193)
* Passing --snifflimit=0 to disable dialect sniffing. (#190)
* Add aarcro to the AUTHORS file.
* Improve performance of csvgrep. (#204)
* Add Matt Dudys to AUTHORS.
* Add support for --skipinitialspace. (#201)
* Add Joakim Lundborg to AUTHORS.
* Add --no-inference option to in2csv and csvsql. (#206)
* Add Federico Scrinzi to AUTHORS file.
* Add --no-header-row to all tools. (#189)
* Fix csvstack blowing up on empty files. (#209)
* Add Chris Rosenthal to AUTHORS file.
* Add --db-schema option to csvsql. (#216)
* Add Shane StClair to AUTHORS file.
* Add --no-inference support to csvsort. (#222)

0.5.0
-----

* Implement geojson support in csvjson. (#159)
* Optimize writing of eight bit codecs. (#175)
* Created csvpy. (#44)
* Support --not-columns for excluding columns. (#137)
* Add Jan Schulz to AUTHORS file.
* Add Windows scripts. (#111, #176)
* csvjoin, csvsql and csvstack will no longer hold open all files. (#178)
* Added Noah Hoffman to AUTHORS.
* Make csvlook output compatible with emacs table markup. (#174)

0.4.4
-----

* Add Derek Wilson to AUTHORS.
* Add Kevin Schaul to AUTHORS.
* Add DBF support to in2csv. (#11, #160)
* Support --zero option for zero-based column indexing. (#144)
* Support mixing nulls and blanks in string columns.
* Add --blanks option to csvsql. (#149)
* Add multi-file (glob) support to csvsql. (#146)
* Add Gregory Temchenko to AUTHORS.
* Add --no-create option to csvsql. (#148)
* Add Anton Ian Sipos to AUTHORS.
* Fix broken pipe errors. (#150)

0.4.3
-----

* Begin CHANGELOG (a bit late, I'll admit).


csvkit/utilities/in2csv.py
#!/usr/bin/env python

import functools
import sys
from io import BytesIO
from os.path import splitext

import agate
import agatedbf  # noqa: F401
import agateexcel  # noqa: F401
import openpyxl
import xlrd

from csvkit import convert
from csvkit.cli import CSVKitUtility
from csvkit.convert.fixed import fixed2csv
from csvkit.convert.geojs import geojson2csv
from csvkit.convert.ods import ods2csv

SUPPORTED_FORMATS = ['csv', 'dbf', 'fixed', 'geojson', 'json', 'ndjson', 'xls', 'xlsx','ods']


class In2CSV(CSVKitUtility):
    description = 'Convert common, but less awesome, tabular data formats to CSV.'
    epilog = 'Some command-line flags only pertain to specific input formats.'
    # The utility handles the input file.
    override_flags = ['f']

    def add_arguments(self):
        self.argparser.add_argument(
            metavar='FILE', nargs='?', dest='input_path',
            help='The CSV file to operate on. If omitted, will accept input as piped data via STDIN.')
        self.argparser.add_argument(
            '-f', '--format', dest='filetype', choices=SUPPORTED_FORMATS,
            help='The format of the input file. If not specified will be inferred from the file type.')
        self.argparser.add_argument(
            '-s', '--schema', dest='schema',
            help='Specify a CSV-formatted schema file for converting fixed-width files. See web documentation.')
        self.argparser.add_argument(
            '-k', '--key', dest='key',
            help='Specify a top-level key to look within for a list of objects to be converted when processing JSON.')
        self.argparser.add_argument(
            '-n', '--names', dest='names_only', action='store_true',
            help='Display sheet names from the input Excel or ODS file.')
        self.argparser.add_argument(
            '--sheet', dest='sheet',
            help='The name of the Excel or ODS sheet to operate on.')
        self.argparser.add_argument(
            '--write-sheets', dest='write_sheets',
            help='The names of the Excel sheets or ODS sheets to write to files, or "-" to write all sheets.')
        self.argparser.add_argument(
            '--use-sheet-names', dest='use_sheet_names', action='store_true',
            help='Use the sheet names as file names when --write-sheets is set.')
        self.argparser.add_argument(
            '--reset-dimensions', dest='reset_dimensions', action='store_true', default=None,
            help='Ignore the sheet dimensions provided by the XLSX file.')
        self.argparser.add_argument(
            '--encoding-xls', dest='encoding_xls',
            help='Specify the encoding of the input XLS file.')
        self.argparser.add_argument(
            '-y', '--snifflimit', dest='sniff_limit', type=int, default=1024,
            help='Limit CSV dialect sniffing to the specified number of bytes. '
                 'Specify "0" to disable sniffing entirely, or "-1" to sniff the entire file.')
        self.argparser.add_argument(
            '-I', '--no-inference', dest='no_inference', action='store_true',
            help='Disable type inference (and --locale, --date-format, --datetime-format, --no-leading-zeroes) '
                 'when parsing CSV input.')

    # This is called only from open_excel_input_file(), but is a separate method to use caching.
    @functools.lru_cache
    def stdin(self):
        return sys.stdin.buffer.read()

    def open_excel_input_file(self, path):
        if not path or path == '-':
            return BytesIO(self.stdin())
        return open(path, 'rb')

    def sheet_names(self, path, filetype):
        input_file = self.open_excel_input_file(path)
        if filetype == 'xls':
            sheet_names = xlrd.open_workbook(file_contents=input_file.read()).sheet_names()
        else:  # 'xlsx'
            sheet_names = openpyxl.load_workbook(input_file, read_only=True, data_only=True).sheetnames
        input_file.close()
        return sheet_names

    def main(self):
        path = self.args.input_path

        # Determine the file type.
        if self.args.filetype:
            filetype = self.args.filetype
        elif self.args.schema:
            filetype = 'fixed'
        elif self.args.key:
            filetype = 'json'
        else:
            if not path or path == '-':
                self.argparser.error('You must specify a format when providing input as piped data via STDIN.')
            filetype = convert.guess_format(path)
            if not filetype:
                self.argparser.error('Unable to automatically determine the format of the input file. Try specifying '
                                     'a format with --format.')

        if self.args.names_only:
            if filetype in ('xls', 'xlsx'):
                sheets = self.sheet_names(path, filetype)
                for sheet in sheets:
                    self.output_file.write(f'{sheet}\n')
            elif filetype in ('ods'):
                sheets = ods2csv(path,display_sheetnames=True)
                for sheet in sheets:
                    self.output_file.write(f'{sheet}\n')
            else:
                self.argparser.error('You cannot use the -n or --names options with non-Excel files.')
            return

        # Set the input file.
        if filetype in ('xls', 'xlsx'):
            self.input_file = self.open_excel_input_file(path)
        else:
            self.input_file = self._open_input_file(path)

        # Set the reader's arguments.
        kwargs = {}
        sniff_limit = self.args.sniff_limit if self.args.sniff_limit != -1 else None

        if self.args.schema:
            schema = self._open_input_file(self.args.schema)
        elif filetype == 'fixed':
            raise ValueError('schema must not be null when format is "fixed"')

        if filetype == 'csv':
            kwargs.update(self.reader_kwargs)
            kwargs['sniff_limit'] = sniff_limit

        if filetype in ('xls', 'xlsx'):
            kwargs['header'] = not self.args.no_header_row

        if filetype not in ('dbf', 'geojson', 'json', 'ndjson'):  # csv, fixed, xls, xlsx
            kwargs['skip_lines'] = self.args.skip_lines

        if filetype != 'dbf':
            kwargs['column_types'] = self.get_column_types()

        if filetype in ('ods'):
            kwargs['sheetname'] = self.args.sheet
            kwargs['write_sheets'] = self.args.write_sheets
            kwargs['display_sheetnames'] = self.args.names_only


        # Convert the file.
        if (
            filetype == 'csv'
            and self.args.no_inference
            and not self.args.no_header_row
            and not self.args.skip_lines
            and sniff_limit == 0
        ):
            reader = agate.csv.reader(self.input_file, **self.reader_kwargs)
            writer = agate.csv.writer(self.output_file, **self.writer_kwargs)
            writer.writerows(reader)
        elif filetype == 'fixed':
            self.output_file.write(fixed2csv(self.input_file, schema, output=self.output_file, **kwargs))
        elif filetype == 'geojson':
            self.output_file.write(geojson2csv(self.input_file, **kwargs))
        elif filetype == 'ods':
            self.output_file.write(ods2csv(self.input_file.name, **kwargs))
        elif filetype in ('csv', 'dbf', 'json', 'ndjson', 'xls', 'xlsx'):
            if filetype == 'csv':
                table = agate.Table.from_csv(self.input_file, **kwargs)
            elif filetype == 'json':
                table = agate.Table.from_json(self.input_file, key=self.args.key, **kwargs)
            elif filetype == 'ndjson':
                table = agate.Table.from_json(self.input_file, key=self.args.key, newline=True, **kwargs)
            elif filetype == 'xls':
                table = agate.Table.from_xls(self.input_file, sheet=self.args.sheet,
                                             encoding_override=self.args.encoding_xls, **kwargs)
            elif filetype == 'xlsx':
                table = agate.Table.from_xlsx(
                    self.input_file, sheet=self.args.sheet, reset_dimensions=self.args.reset_dimensions, **kwargs
                )
            elif filetype == 'dbf':
                if not hasattr(self.input_file, 'name'):
                    raise ValueError('DBF files can not be converted from stdin. You must pass a filename.')
                table = agate.Table.from_dbf(self.input_file.name, **kwargs)
            table.to_csv(self.output_file, **self.writer_kwargs)

        if self.args.write_sheets:
            if filetype in ('xls', 'xlsx'):
                # Close and re-open the file, as the file object has been mutated or closed.
                self.input_file.close()

                self.input_file = self.open_excel_input_file(path)

                if self.args.write_sheets == '-':
                    sheets = self.sheet_names(path, filetype)
                else:
                    sheets = [int(sheet) if sheet.isdigit() else sheet for sheet in self.args.write_sheets.split(',')]

                if filetype == 'xls':
                    tables = agate.Table.from_xls(self.input_file, sheet=sheets,
                                                encoding_override=self.args.encoding_xls, **kwargs)
                elif filetype == 'xlsx':
                    tables = agate.Table.from_xlsx(
                        self.input_file, sheet=sheets, reset_dimensions=self.args.reset_dimensions, **kwargs
                    )

                if not path or path == '-':
                    base = 'stdin'
                else:
                    base = splitext(self.input_file.name)[0]
                for i, (sheet_name, table) in enumerate(tables.items()):
                    if self.args.use_sheet_names:
                        filename = '%s_%s.csv' % (base, sheet_name)
                    else:
                        filename = '%s_%d.csv' % (base, i)
                    with open(filename, 'w') as f:
                        table.to_csv(f, **self.writer_kwargs)

        self.input_file.close()

        if self.args.schema:
            schema.close()


def launch_new_instance():
    utility = In2CSV()
    utility.run()


if __name__ == '__main__':
    launch_new_instance()
